FindAllShortestPaths.cpp

```diff
@@ -0,0 +1,198 @@
+            }
+        }
+    }
+
+    return Parents;
+}
+
+// this function will recover all paths in `answer` using `Parents` matrix
+void RecoverPathsHelper(const std::vector<std::vector<std::pair<int, int>>>& Parents, int toVertex, int fromVertex,
+    std::vector<std::vector<int>>& answer, std::vector<int> onePath)
+{
+    // this is modified DFS algorithm which will recover all paths
+    std::stack<std::pair<int, int>> stack;
+
+    // pair< vertex in stack , index of element from `Parents[vertex]` >
+    stack.push( std::make_pair( fromVertex, 0 ) );
+
+    while (!stack.empty())
+    {
+        std::pair<int, int> s = stack.top();
+
+        onePath.push_back(s.first);
+
+        // if `s.first` is equal to toVertex then we need to save the answer and go back to see if there is more paths
+        if (s.first == toVertex)
+        {
+            std::vector<int> reversedOnePath(onePath.rbegin(), onePath.rend());
+            answer.push_back(reversedOnePath);
+        }
+
+        bool s_has_neighbors = false;
+
+        /*
+         we get from `Parents[s.first]` next index that we need. For example if `s` is pair<6, 2> then we need to push in stack Parents[6][2] and change
+         `s` to pair<6, 3>
+        */
+        if (s.second < Parents[s.first].size())
+        {
+            s_has_neighbors = true;
+            std::pair<int, int> helper = Parents[s.first][s.second];
+            helper.second = 0;
+            stack.push(helper);
+        }
+
+        if (!s_has_neighbors)
+        {
+            stack.pop();
+            if (stack.empty())
+            {
+                return;
+            }
+            s = stack.top();
+            ++s.second;
+            stack.pop();
+            stack.push(s);
+            onePath.resize(onePath.size() - 2);
+        }
+    }
+}
+
+std::vector<std::vector<int>> RecoverPaths(const std::vector<std::vector<std::pair<int, int>>>& Parents, int toVertex, int fromVertex)
+{
+    std::vector<std::vector<int>> answer;
+    std::vector<int> onePath;
+    RecoverPathsHelper(Parents, toVertex, fromVertex, answer, onePath);
+
+    return answer;
+}
+
+adjacency_list ReadGraph()
+{
+    int vertices, edges;
+    std::cin >> vertices >> edges;
+
+    adjacency_list Graph(vertices);
+
+    for (int i = 0; i < edges; ++i)
+    {
+        int a, b, weight;
+        std::cin >> a >> b >> weight;
+
+        Graph[a].push_back(std::make_pair(b, weight));
+        Graph[b].push_back(std::make_pair(a, weight));
+    }
+
+    return Graph;
+}
+
+void PrintGraph(const adjacency_list& Graph)
+{
+    int index = 0;
+    for (auto x : Graph)
+    {
+        std::cout << index << ": ";
+        for (auto y : x)
+        {
+            std::cout << y.first << "(" << y.second << ")  ";
+        }
+        std::cout << "\n";
+        ++index;
+    }
+}
+
+int main()
+{
+    adjacency_list Graph = ReadGraph();
+
+    int fromVertex, toVertex;
+
+    std::cin >> fromVertex >> toVertex;
```

Comment:
Where is the validation checks? what if the user have inserted invalid values? `-4` for example

Reply:
done